src/lib.rs

```diff
@@ -87,21 +93,42 @@ fn parse_sealed(item: syn::Item) -> syn::Result<proc_macro2::TokenStream> {
 }
 
 // Care for https://gist.github.com/Koxiaet/8c05ebd4e0e9347eb05f265dfb7252e1#procedural-macros-support-renaming-the-crate
-fn parse_sealed_struct(strct: syn::ItemStruct) -> syn::Result<proc_macro2::TokenStream> {
-    let ident = &strct.ident;
-    Ok(quote!(
-        #strct
-        impl private::Sealed for #ident {}
-    ))
+fn parse_sealed_trait(mut item_trait: syn::ItemTrait) -> TokenStream2 {
+    let trait_ident = &item_trait.ident.unraw();
+    let trait_generics = &item_trait.generics;
+    let seal = seal_name(trait_ident);
+    item_trait
+        .supertraits
+        .push(parse_quote!(#seal::Sealed #trait_generics));
+    quote!(
+        pub(crate) mod #seal {
+            pub trait Sealed #trait_generics {}
```

Comment:
Also, an important thing is to `use super::*` in the generate module. Since, well, we have trait bounds:

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9cb505bbdd163e197125a37eb50ca259)

Reply:
Indeed! That being said, `use super::*;` breaks https://rust-lang.github.io/api-guidelines/macros.html#item-macros-work-anywhere-that-items-are-allowed-c-anywhere, so if it can be avoided it will be a win. In this instance, it can be avoided by removing the bounds, so it's mostly a matter of adding:

```rust
let trait_generics = trait_generics.split_for_impl().0;
```

before the `quote!` invocation

Reply:
@jmg-duarte @danielhenrymantilla if so, we need to consider implicit `Sized`. So, something like this would work:
```rust
#[sealed]
pub MyTrait<T: ?Sized>: MyBounds {}

// As described above will desugar as:
pub MyTrait<T: ?Sized>: MyBounds + private::Sealed<T> {}
mod private {
    pub Sealed<T> {}
}

// Which breaks the original trait:
impl private::Sealed<dyn Error> for MyType {} // compler error

// If we remove bounds, we should allways apply `?Sized` to type params,
// to be transparent asap:
pub MyTrait<T: ?Sized>: MyBounds + private::Sealed<T> {}
mod private {
    pub Sealed<T: ?Sized> {}
}
```

Reply:
Ah, true! The `Sealed` trait could then be made generic over `?Sized` types to avoid this; although I do admit that at this point using `use super::*;` (so as to forward _all_ the bounds) at the cost of not supporting being applied _inside_ a function's body may be the simpler way to go. I don't have strong preferences anymore.

Reply:
@jmg-duarte @danielhenrymantilla I'd prefer to start with a simpler `use super::*` solution and mess around only if there will be any real cases working badly this way.